hologram_dashboard/fetcher/hologram_data_source_availability_fetcher.cc

```diff
@@ -34,4 +34,53 @@ void HologramDataSourceAvailabilityFetcher::
                                                 &hologram_configs_));
 }
 
+void HologramDataSourceAvailabilityFetcher::UpdateProto(std::string system, 
+    std::time_t time, DataSource data_source, StatusType status) {
+    std::unordered_map<DataSource,HologramDataAvailability>& 
+        data_to_availability_map = data_sources_availability_map_[system];
+    HologramDataAvailability& hda = data_to_availability_map[data_source];
+    DataSourceDetail* latest_status = hda.mutable_latest_status();
+    if (!hda.has_data_source()) {
+        hda.set_data_source(data_source);
+    }
+    latest_status->set_date((int) time);
+    latest_status->set_status(status);
+    UpdateHistory(hda, time, status);
+}
+
+void HologramDataSourceAvailabilityFetcher::UpdateHistory(
+    HologramDataAvailability& hda, std::time_t time, StatusType status) {
+    if (hda.history_size() != 0) {
+        // Acquire the latest history
+        auto latest_history = hda.history(hda.history_size() - 1);
+        time_t history_time = latest_history.date();
+        struct tm time_tm;
+        struct tm history_time_tm;
+        // Have to use localtime_r since localtime will overwrite each other
+        localtime_r(&time, &time_tm);
+        localtime_r(&history_time, &history_time_tm);
+        // Note here we ignore the case in which the given time is lower than
+        // the latest status because user should not be able to go back in time
+        // and check details of the past.
+        if (time_tm.tm_year == history_time_tm.tm_year &&
+            time_tm.tm_mon == history_time_tm.tm_mon &&
+            time_tm.tm_mday == history_time_tm.tm_mday) {
+            // If it's the same day just update the status.
+            latest_history.set_date((int) time);
+            latest_history.set_status(status);
+            return;
+        }
+    }
+    // Add a history if either its empty or it is later than latest history.
+    DataSourceDetail* detail = hda.add_history();
+    detail->set_date((int) time);
+    detail->set_status(status);
+
+    if (hda.history_size() > 7) {
+        // Remove the oldest history
+        auto history_list = hda.mutable_history();
+        history_list->DeleteSubrange(0, 1);
+    }
```

Comment:
Don't remove history.

Reply:
Don't we want to cap the number of history at 7(or 35)?

Reply:
ping